### PR TITLE
Propagate fee info to `BlockHeader::new`

### PR DESCRIPTION
**Pull request by QuentinI**
Partially addresses #2962 

### This PR: 
- Adds fee information to `Header::new`
- Propagates such information from transaction task as received from the builder

### This PR does not: 
Implement any changes to VID described in #2962, that will be implemented as a separate PR. 

### Key places to review: 
All non-testing code, changes in testing code are mostly trivial

### CODE REVIEW

File: crates/testing/src/task_helpers.rs [hunk 1]
```diff
@@ -212,141 +208,6 @@ pub fn build_assembled_sig<
     real_qc_sig
 }
 
-/// build a quorum proposal and signature
-#[allow(clippy::too_many_lines)]
-async fn build_quorum_proposal_and_signature(
```
Comment (move47) [inline crates/testing/src/task_helpers.rs:217]: I am not sure but is this removal intentional?

File: crates/task-impls/src/consensus.rs [hunk 2]
```diff
@@ -51,11 +51,13 @@ use crate::{
     },
 };
 /// Alias for the block payload commitment and the associated metadata.
-pub struct CommitmentAndMetadata<PAYLOAD: BlockPayload> {
+pub struct CommitmentAndMetadata<TYPES: NodeType> {
```
Comment (jparr721) [inline crates/task-impls/src/consensus.rs:54]: Might be worth also updating this doc comment to explain the purpose of the fee data.

File: crates/builder-api/src/block_info.rs [hunk 3]
```diff
@@ -41,5 +41,4 @@ pub struct AvailableBlockHeaderInput<TYPES: NodeType> {
     pub message_signature:
         <<TYPES as NodeType>::BuilderSignatureKey as BuilderSignatureKey>::BuilderSignature,
     pub sender: <TYPES as NodeType>::BuilderSignatureKey,
-    pub _phantom: PhantomData<TYPES>,
```
Comment (jparr721) [inline crates/builder-api/src/block_info.rs:44]: Nice.

File: crates/types/src/traits/block_contents.rs [hunk 4]
```diff
@@ -145,6 +146,15 @@ pub fn precompute_vid_commitment(
 /// do dispersal for the genesis block. For simplicity and performance, we use 1.
 pub const GENESIS_VID_NUM_STORAGE_NODES: usize = 1;
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Information about builder fee for proposed block
+pub struct FeeData<TYPES: NodeType> {
```
Comment (jparr721) [inline crates/types/src/traits/block_contents.rs:151]: Could we perhaps rename this? It's not so much Fee data as it is just the builder fee for a proposed block in my mind at least. Open to discussion, but, perhaps something like `BuilderFee` would be more descriptive?

File: crates/types/src/data.rs [hunk 5]
```diff
@@ -692,4 +698,29 @@ pub mod null_block {
...
+            <TYPES::BuilderSignatureKey as BuilderSignatureKey>::generated_from_seed_indexed(
+                [0_u8; 32], 0,
+            );
+
+        let (null_block, null_block_metadata) =
+            <TYPES::BlockPayload as BlockPayload>::from_transactions([]).ok()?;
+
+        match TYPES::BuilderSignatureKey::sign_fee(
+            &priv_key,
+            0,
```
Comment (jparr721) [inline crates/types/src/data.rs:715]: Could this either not be a magic number or be something like `0 /* what this value is */`?

File: crates/types/src/data.rs [hunk 6]
```diff
@@ -692,4 +698,29 @@ pub mod null_block {
...
+            &null_block.builder_commitment(&null_block_metadata),
+            &commitment(num_storage_nodes)?,
+        ) {
+            Ok(sig) => Some(FeeData {
+                fee_amount: 0,
```
Comment (jparr721) [inline crates/types/src/data.rs:720]: Could we get a comment explaining why the amount is 0 here? It's hard to understand if I didn't already have context.

File: crates/task-impls/src/events.rs [hunk 7]
```diff
@@ -106,12 +106,14 @@ pub enum HotShotEvent<TYPES: NodeType> {
...
         <TYPES::BlockPayload as BlockPayload>::Metadata,
         TYPES::Time,
+        FeeData<TYPES>,
     ),
     /// Event when the transactions task has sequenced transactions. Contains the encoded transactions, the metadata, and the view number
     BlockRecv(
         Vec<u8>,
         <TYPES::BlockPayload as BlockPayload>::Metadata,
         TYPES::Time,
+        FeeData<TYPES>,
```
Comment (jparr721) [inline crates/task-impls/src/events.rs:116]: I don't think we should change anything right now, but I just want to note that these gradually-longer event types might start to become painful as time goes on, but I don't have a good solution yet.

File: crates/task-impls/src/consensus.rs [hunk 8]
```diff
@@ -219,13 +221,15 @@ async fn create_and_send_proposal<TYPES: NodeType>(
     upgrade_cert: Option<UpgradeCertificate<TYPES>>,
     proposal_cert: Option<ViewChangeEvidence<TYPES>>,
     round_start_delay: u64,
+    fee_data: FeeData<TYPES>,
```
Comment (jparr721) [inline crates/task-impls/src/consensus.rs:224]: This parameter list is getting pretty painful. Is it possible to perhaps just build `CommitmentAndMetadata` and pass that in directly instead of adding yet-another parameter to this list? It looks like the caller will have access to everything.

File: crates/task-impls/src/quorum_proposal.rs [hunk 9]
```diff
@@ -183,6 +182,7 @@ impl<TYPES: NodeType> ProposalDependencyHandle<TYPES> {
             &parent_leaf,
             commit_and_metadata.commitment,
             commit_and_metadata.metadata.clone(),
+            commit_and_metadata.fee_data.clone(),
```
Comment (jparr721) [inline crates/task-impls/src/quorum_proposal.rs:185]: If we're moving `commit_and_metadata` do we need to clone this?

File: crates/testing/src/block_builder.rs [hunk 10]
```diff
@@ -518,50 +518,21 @@ fn build_block<TYPES: NodeType>(
...
-        block_info.extend_from_slice(commitment.as_ref());
-
-        match TYPES::BuilderSignatureKey::sign_builder_message(&priv_key, &block_info) {
-            Ok(sig) => sig,
-            Err(e) => {
-                panic!("Failed to sign block: {}", e);
-            }
-        }
-    };
+    let signature_over_block_info =
```
Comment (jparr721) [inline crates/testing/src/block_builder.rs:521]: This is very nice cleanup and much easier to parse, thanks for doing this.

**Review by jparr721 — CHANGES_REQUESTED** [review]
Open to discussion here, but I think the name change might be worth doing since it's pretty easy, and also maybe some more comments where I've indicated.

File: crates/task-impls/src/events.rs (hunk 7 above)
Comment (QuentinI) [reply crates/task-impls/src/events.rs:116]: Agreed. I'd go further and say that we're overusing tuples, especially in enum variants, and propose something drastic like banning tuples with more than two fields or such. I'm going to raise this in today's Consensus meeting.

File: crates/task-impls/src/quorum_proposal.rs (hunk 9 above)
Comment (QuentinI) [reply crates/task-impls/src/quorum_proposal.rs:185]: We don't, thanks! Strange that clippy didn't notice.

**Review by jparr721 — APPROVED** [review]
Great work

**Review by move47 — APPROVED** [review]

**Review by jparr721 — DISMISSED** [review]

**Review by jparr721 — APPROVED** [review]
